### ISSUE TITLE
Windows: Too long paths when building from source

### ISSUE THREAD

**Issue by thopiekar** [cython/cython#2515]
Here an example:
```
building 'Cython.Compiler.FlowControl' extension
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\x86_amd64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -ID:\Windows\cdk\cura-build-environment\install\opt\Python\include -ID:\Windows\cdk\cura-build-environment\install\opt\Python\include "-IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" "-IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17134.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17134.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17134.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17134.0\winrt" -ID:\Windows\cdk\cura-build-environment\install\include "-IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" "-IC:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17134.0\ucrt" "-IC:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17134.0\shared" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17134.0\um" "-IC:\Program Files (x86)\Windows Kits\10\include\10.0.17134.0\winrt" /TcD:\Windows\cdk\cura-build-environment\build\cython-prefix\src\cython\Cython\Compiler\FlowControl.c /Fobuild\temp.win-amd64-3.5\Release\Windows\cdk\cura-build-environment\build\cython-prefix\src\cython\Cython\Compiler\FlowControl.obj
FlowControl.c
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\x86_amd64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:D:\Windows\cdk\cura-build-environment\install\opt\Python\libs /LIBPATH:D:\Windows\cdk\cura-build-environment\install\opt\Python\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\LIB\amd64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\LIB\amd64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17134.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17134.0\um\x64" /LIBPATH:D:\Windows\cdk\cura-build-environment\install\lib /LIBPATH:D:\Windows\cdk\cura-build-environment\install\bin "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\LIB\amd64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\LIB\amd64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17134.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.17134.0\um\x64" /EXPORT:PyInit_FlowControl build\temp.win-amd64-3.5\Release\Windows\cdk\cura-build-environment\build\cython-prefix\src\cython\Cython\Compiler\FlowControl.obj /OUT:build\lib.win-amd64-3.5\Cython\Compiler\FlowControl.cp35-win_amd64.pyd /IMPLIB:build\temp.win-amd64-3.5\Release\Windows\cdk\cura-build-environment\build\cython-prefix\src\cython\Cython\Compiler\FlowControl.cp35-win_amd64.lib
FlowControl.obj : warning LNK4197: export 'PyInit_FlowControl' specified multiple times; using first specification
   Creating library build\temp.win-amd64-3.5\Release\Windows\cdk\cura-build-environment\build\cython-prefix\src\cython\Cython\Compiler\FlowControl.cp35-win_amd64.lib and object build\temp.win-amd64-3.5\Release\Windows\cdk\cura-build-environment\build\cython-prefix\src\cython\Cython\Compiler\FlowControl.cp35-win_amd64.exp
Generating code
Finished generating code
```
So what you can see is that
```object build\temp.win-amd64-3.5\Release\Windows\cdk\cura-build-environment\build\cython-prefix\src\cython\Cython\```
is being used to build all binaries. However, this makes the path very long for no reason. What I expect is to use
```build\temp.win-amd64-3.5\Release\Cython\```
, because ```\Windows\cdk\cura-build-environment\build\cython-prefix\src\cython``` is the location (without drive letter) where the sources are stored at build time.
So it looks like someone is running os.path.join(TEMP_DIR, SOURCE_FILES) at some point, but doing a os.path.realpath on SOURCE_FILES and not os.path.relpath.

I tried to fix it on my own instead of opening a issue report, but couldn't find it in ```setup.py```.

**Comment by thopiekar** [comment]
Oh, and the problem is quite serious, because you will get errors on Windows because of too long paths.
At least, I think that is the reason why my builds fail here. However, somewhere else on a lower level it just works.

**Comment by jasonhildebrand** [comment]
It's a legacy server, running Windows 2008.  I'm using Python 3.8.  In response to your message, I updated pip and setuptools to the latest versions (22.1.2 and 63.2.0, respectively).  I tried re-installing but got the same error.

I am trying to build cyanodbc, which requires Cython.  Digging in, I saw it was trying to install a pinned version of Cython - 0.28.2.  I have changed this to a more recent version (0.29.30), and it installed successfully using a wheel.
